src/main/java/fi/aalto/cs/apluscourses/utils/StateMonitor.java

```diff
@@ -0,0 +1,125 @@
+package fi.aalto.cs.apluscourses.utils;
+
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * A synchronized implementation of incrementally changing state.  Different states are represented
+ * as different integers, semantics of which can be chosen by the client code with the following
+ * rules:
+ * <ul>
+ *   <li>The initial state is {@code INITIAL}</li>.
+ *   <li>All the states greater than or equal to {@code INITIAL} are considered non-error states.
+ *   All the states less than or equal to {@code ERROR} are considered error states.</li>
+ *   <li>Normally, the state can only increase, that is, on a single change, a new state can never
+ *   be less than the old state, except when either the old state or the new state is an error
+ *   state.</li>
+ * </ul>
+ *
+ * <p>The {@link StateListener}, given in the constructor is notified of the changes in state.</p>
+ */
+public class StateMonitor {
+  private int state = INITIAL;
```

Comment:
Nit: maybe these "state"s could be "stateCode"s?

Reply:
Related: https://github.com/Aalto-LeTech/intellij-plugin/pull/57#discussion_r379926959